File: bench/prongB/json_parse.go
// json-parse prong B — Go (GC). Recursive-descent into a pointer Json tree,
// walk-checksum, discard (garbage collected). K passes over a stdin doc.
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

type Json struct {
	tag  int // 0 null 1 bool 2 num 3 str 4 arr 5 obj
	num  int64
	str  string
	keys []string
	vals []*Json
}

var s []byte
var n, p int

func skipWs() {
	for p < n {
		c := s[p]
		if c == 32 || c == 9 || c == 10 || c == 13 {
			p++
		} else {
			break
		}
	}
}
func parseString() string {
	p++
	var r []byte
	for p < n && s[p] != '"' {
		var c byte
		if s[p] == '\\' {
			p++
			e := s[p]
			switch e {
			case 'n':
				c = '\n'
			case 't':
				c = '\t'
			case '"':
				c = '"'
			case '\\':
				c = '\\'
			default:
				c = e
			}
		} else {
			c = s[p]
		}
		r = append(r, c)
		p++
	}
	p++
	return string(r)
}
func parseNumber() int64 {
	neg := false
	if s[p] == '-' {
		neg = true
		p++
	}
	var v int64 = 0
	for p < n && s[p] >= '0' && s[p] <= '9' {
		v = v*10 + int64(s[p]-'0')
		p++
	}
	if neg {
		return -v
	}
	return v
}
func parseArray() *Json {
	p++
	j := &Json{tag: 4}
	skipWs()
	for p < n && s[p] != ']' {
		j.vals = append(j.vals, parseValue())
		skipWs()
		if p < n && s[p] == ',' {
			p++
			skipWs()
		}
	}
	p++
	return j
}
func parseObject() *Json {
	p++
	j := &Json{tag: 5}
	skipWs()
	for p < n && s[p] != '}' {
		k := parseString()
		skipWs()
		p++
		j.keys = append(j.keys, k)
		j.vals = append(j.vals, parseValue())
		skipWs()
		if p < n && s[p] == ',' {
			p++
			skipWs()
		}
	}
	p++
	return j
}
func parseValue() *Json {
	skipWs()
	c := s[p]
	switch {
	case c == '{':
		return parseObject()
	case c == '[':
		return parseArray()
	case c == '"':
		return &Json{tag: 3, str: parseString()}
	case c == 't':
		p += 4
		return &Json{tag: 1, num: 1}
	case c == 'f':
		p += 5
		return &Json{tag: 1, num: 0}
	case c == 'n':
		p += 4
		return &Json{tag: 0}
	default:
		return &Json{tag: 2, num: parseNumber()}
	}
}
func walk(j *Json) int64 {
	switch j.tag {
	case 2:
		return j.num + 1
	case 3:
		return int64(len(j.str)) + 1
	case 1:
		return 1
	case 0:
		return 1
	case 4:
		var t int64 = 1
		for _, x := range j.vals {
			t += walk(x)
		}
		return t
	case 5:
		var t int64 = 1
		for _, k := range j.keys {
			t += int64(len(k))
		}
		for _, v := range j.vals {
			t += walk(v)
		}
		return t
	}
	return 0
}
func main() {
	buf, _ := io.ReadAll(bufio.NewReader(os.Stdin))
	s = buf
	n = len(buf)
	var acc int64 = 0
	for k := 0; k < 30; k++ {
		p = 0
		j := parseValue()
		acc += walk(j)
	}
	fmt.Printf("checksum=%d\n", acc)
}

File: bench/prongB/json_parse.rs
// json-parse prong B — Rust. Recursive-descent parse into an owned `Json` enum,
// walk-checksum, drop (RAII). K parse-and-discard passes over a stdin doc.
use std::io::Read;

enum Json { Null, Bool(bool), Num(i64), Str(String), Arr(Vec<Json>), Obj(Vec<String>, Vec<Json>) }

struct P<'a> { s: &'a [u8], pos: usize, n: usize }
impl<'a> P<'a> {
    fn skip_ws(&mut self) { while self.pos < self.n { let c = self.s[self.pos]; if c==32||c==9||c==10||c==13 { self.pos += 1; } else { break; } } }
    fn parse_string(&mut self) -> String {
        self.pos += 1; let mut r = String::new();
        while self.pos < self.n && self.s[self.pos] != b'"' {
            let c = if self.s[self.pos] == b'\\' { self.pos += 1; let e = self.s[self.pos];
                match e { b'n'=>b'\n', b't'=>b'\t', b'"'=>b'"', b'\\'=>b'\\', _=>e } } else { self.s[self.pos] };
            r.push(c as char); self.pos += 1;
        }
        self.pos += 1; r
    }
    fn parse_number(&mut self) -> i64 {
        let mut neg = false; if self.s[self.pos] == b'-' { neg = true; self.pos += 1; }
        let mut v: i64 = 0;
        while self.pos < self.n && self.s[self.pos] >= b'0' && self.s[self.pos] <= b'9' { v = v*10 + (self.s[self.pos]-b'0') as i64; self.pos += 1; }
        if neg { -v } else { v }
    }
    fn parse_array(&mut self) -> Json {
        self.pos += 1; let mut xs = Vec::new(); self.skip_ws();
        while self.pos < self.n && self.s[self.pos] != b']' {
            xs.push(self.parse_value()); self.skip_ws();
            if self.pos < self.n && self.s[self.pos] == b',' { self.pos += 1; self.skip_ws(); }
        }
        self.pos += 1; Json::Arr(xs)
    }
    fn parse_object(&mut self) -> Json {
        self.pos += 1; let mut ks = Vec::new(); let mut vs = Vec::new(); self.skip_ws();
        while self.pos < self.n && self.s[self.pos] != b'}' {
            let k = self.parse_string(); self.skip_ws(); self.pos += 1;
            ks.push(k); vs.push(self.parse_value()); self.skip_ws();
            if self.pos < self.n && self.s[self.pos] == b',' { self.pos += 1; self.skip_ws(); }
        }
        self.pos += 1; Json::Obj(ks, vs)
    }
    fn parse_value(&mut self) -> Json {
        self.skip_ws();
        match self.s[self.pos] {
            b'{' => self.parse_object(),
            b'[' => self.parse_array(),
            b'"' => Json::Str(self.parse_string()),
            b't' => { self.pos += 4; Json::Bool(true) },
            b'f' => { self.pos += 5; Json::Bool(false) },
            b'n' => { self.pos += 4; Json::Null },
            _ => Json::Num(self.parse_number()),
        }
    }
}
fn walk(j: &Json) -> i64 {
    match j {
        Json::Num(x) => x + 1,
        Json::Str(s) => s.len() as i64 + 1,
        Json::Bool(_) => 1,
        Json::Null => 1,
        Json::Arr(xs) => { let mut t = 1i64; for x in xs { t += walk(x); } t },
        Json::Obj(ks, vs) => { let mut t = 1i64; for k in ks { t += k.len() as i64; } for v in vs { t += walk(v); } t },
    }
}
fn main() {
    let mut buf = Vec::new(); std::io::stdin().read_to_end(&mut buf).unwrap();
    let mut acc: i64 = 0;
    for _ in 0..30 {
        let mut p = P { s: &buf, pos: 0, n: buf.len() };
        let j = p.parse_value();
        acc += walk(&j);
    }
    println!("checksum={}", acc);
}
